designcompose/src/main/java/com/android/designcompose/squoosh/SquooshTreeBuilder.kt

```diff
@@ -289,8 +295,23 @@ internal fun resolveVariantsRecursively(
                 hasSupportedInteraction ||
                     r.trigger is Trigger.OnClick ||
                     r.trigger is Trigger.OnPress
+            if (r.trigger is Trigger.OnKeyDown) {
+                // Register to be a listener for key reactions on this node
+                val keyTrigger = r.trigger as Trigger.OnKeyDown
+                val keyEvent = DesignKeyEvent.fromJsKeyCodes(keyTrigger.key_codes)
+                val keyAction =
+                    KeyAction(
+                        interactionState,
+                        r.action,
+                        findTargetInstanceId(document, parentComps, r.action),
+                        customizations.getKey(),
+                        null
+                    )
+                keyTracker.addListener(keyEvent, keyAction)
```

Comment:
I think because we build a new tree every iteration (or two new trees if there's an animation) this will cause the list of listeners to get huge. Another way to do it might be to have one listener in SquooshRoot that iterates the current SquooshResolvedNode tree and sees if any node has a key event handler. OR make resolveVariantsRecursively generate a list of `(keyEvent, keyAction)`, and the SquooshRoot key event handler iterates the most recent version of that list?

Reply:
That's what I basically did originally. I had a global list and cleared out the list every iteration and built a new one in the `resolveVariantsRecursively`. However, since it was global the design switcher's root node cleared out everything when it shouldn't have. Then I made an instance of `KeyTracker` per document. I'll make some improvements here.

I don't think this list will get huge because it stores using a `HashSet`, but I don't think nodes that change under root and no longer have key event interactions will update. I'll fix that.